Integrate paging for `Topic Screen`

yesshreyes opened the pull request.

yesshreyes committed:
Integrate multiplatform paging for `Topic Screen`

Key changes:
- Added `TopicPagingSource` to handle fetching paged data.
- Modified `TopicViewModel` to use `Pager` for fetching and managing paged data.
- Updated `TopicList` and `TopicScreenUI` to consume `LazyPagingItems`.
- Added new dependencies for multiplatform paging (`cashapp-paging-common`, `cashapp-paging-compose-common`).
- Updated `TopicRepository` with `getTopicsPage` function to support paged fetching.
- Updated previews to reflect the new paged data structure.

yesshreyes committed:
Integrate multiplatform paging for `Topic Screen`

Key changes:
- Added `TopicPagingSource` to handle fetching paged data.
- Modified `TopicViewModel` to use `Pager` for fetching and managing paged data.
- Updated `TopicList` and `TopicScreenUI` to consume `LazyPagingItems`.
- Added new dependencies for multiplatform paging (`cashapp-paging-common`, `cashapp-paging-compose-common`).
- Updated `TopicRepository` with `getTopicsPage` function to support paged fetching.
- Updated previews to reflect the new paged data structure.

RajashekarRaju reviewed (changes requested):
As discussed in slack. Please review the PR yourself and improve quality overall, remove redundant code after switching to new implementation.

yesshreyes committed:
- Update `TopicViewModel` to use `PagingData<TopicUi>` directly, removing the intermediate `UiState` and `fetchTopicList` logic.
- Modify `TopicRepository.getTopicsPage` to return `List<TopicUi>`.
- Make `TopicRepository.getTopics` private as it's now an internal detail.
- Update UI components (`TopicList`, `TopicCard`, `TopicScreenUI`) and preview utils to reflect these changes.
- Adjust `TopicPagingSource` to work with `TopicUi`.
- Rename `ItemTopic` to `TopicUi` and move it to a separate file.

RajashekarRaju commented on composeApp/build.gradle.kts:
```diff
@@ -88,6 +88,7 @@ kotlin {
             implementation(libs.generativeai)
             implementation(compose.uiTooling)
             implementation(libs.ktor.client.okhttp)
+            implementation("androidx.paging:paging-compose:3.3.0-alpha02")
```
Why do we need this still?

RajashekarRaju commented on composeApp/src/commonMain/kotlin/com/developersbreach/kotlindictionarymultiplatform/ui/screens/topic/TopicUi.kt:
```diff
@@ -0,0 +1,7 @@
+package com.developersbreach.kotlindictionarymultiplatform.ui.screens.topic
+
+data class TopicUi(
```
This is very confusing and misleading, this should be renamed to either `Toipc` or `ItemTopic`

RajashekarRaju commented on composeApp/src/commonMain/kotlin/com/developersbreach/kotlindictionarymultiplatform/ui/screens/topic/TopicScreenUI.kt:
```diff
@@ -9,9 +9,11 @@ import androidx.compose.ui.Modifier
 import androidx.compose.ui.unit.dp
 import com.developersbreach.designsystem.components.KdScaffold
 
```
Keeping these clean wouldn't hurt. Keep it consistent.
Also why was this not caught by `ktlint`?

yesshreyes replied:
It is being flagged if there are more than one consecutive empty lines.

RajashekarRaju commented on composeApp/src/commonMain/kotlin/com/developersbreach/kotlindictionarymultiplatform/ui/screens/topic/TopicPagingSource.kt:
```diff
@@ -0,0 +1,34 @@
+    override fun getRefreshKey(
+        state: PagingState<Int, TopicUi>,
+    ): Int {
+        return 1
```
This is not common in any paging implementation, why we need to return 1 always?
Shouldn't you use the paging state’s anchorPosition, not always 1?

RajashekarRaju commented on composeApp/src/commonMain/kotlin/com/developersbreach/kotlindictionarymultiplatform/ui/screens/topic/TopicViewModel.kt:
```diff
@@ -2,68 +2,43 @@ package com.developersbreach.kotlindictionarymultiplatform.ui.screens.topic
-        viewModelScope.launch {
-            fetchTopicList()
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val pagingDataFlow: Flow<PagingData<TopicUi>> = searchQuery
```
As we discussed previosly, you don't need to tell what sort of type it is for all variable names, for instance this could be called just "topics".

RajashekarRaju commented on composeApp/src/commonMain/kotlin/com/developersbreach/kotlindictionarymultiplatform/ui/screens/topic/TopicPagingSource.kt:
```diff
@@ -0,0 +1,34 @@
+import app.cash.paging.PagingState
+import com.developersbreach.kotlindictionarymultiplatform.data.topic.repository.TopicRepository
+
+class TopicPagingSource(
```
This should be located in 
```
package com.developersbreach.kotlindictionarymultiplatform.data.topic.repository
```

yesshreyes committed:
- Rename `TopicUi` to `Topic` for better clarity.
- Move `TopicPagingSource` to the data layer.
- Update `TopicViewModel`, `TopicList`, `TopicCard`, and `TopicScreen` to use the new `Topic` data class.
- Adjust `TopicRepository` to return `Topic` objects.
- Enhance `TopicPagingSource`'s `getRefreshKey` logic.
- Update `PreviewUtils` to use the renamed `Topic` class.
- Remove unused Android Paging dependency.

RajashekarRaju committed:
Make corrections to model classes and variables

RajashekarRaju reviewed (commented):
Review my commit, do you believe this is what you needed?
https://github.com/DevelopersBreach/kotlin-dictionary-multiplatform/pull/25/commits/24f240ea63a94adce0752f92e21cf5b85fc62434

yesshreyes commented:
> Review my commit, do you believe this is what you needed? [24f240e](https://github.com/DevelopersBreach/kotlin-dictionary-multiplatform/commit/24f240ea63a94adce0752f92e21cf5b85fc62434)

Reviewed.
